React rewrite of Babel REPL

bvaughn opened the pull request:
Relates to #1287; cc @hzoo, @Daniel15, @existentialism, @CompuIves

Preview available at [bvaughn.github.io/babel-repl](https://bvaughn.github.io/babel-repl/)

### High-level features
* Comlpetely rewritten with React for easier future maintanence.
* Fully Flow-typed.
* Large plugins (eg Babili, Prettier) load on-demand, only if enabled.
  * Loading doesn't block compilation; once a plugin loads, code will be auto-recompiled.
  * This framework can be extended for third party plugins (loaded directly from NPM).
* Large libraries like Codemirror and Babel are loaded from a CDN to better share with other sites (eg [prettier.io](https://prettier.io/)).
* Mobile-friendly layout.

### Structure

```sh
.
├── _site
│   ├── repl  # current REPL source
│   └── repl2 # new REPL build output
└── js
    └── repl  # new REPL source
```

### Open questions
* Are we okay with using create-react-app for the React portions of the site or should I trim down the Webpack+Babel configuration.

### Follow-up items in the ["New REPL" milestone](https://github.com/babel/website/milestone/1)
* #1309: Configure `make build` to run `yarn build` (in `src`) and `make watch` to run `yarn start`.
* #1310: Run `yarn flow` and `yarn prettier` as part of precommit hook (or a CI check).
* #1311: Replace `babili-standalone` with `babel-minifi-standalone` once the new package has been published.
* #1312: Don't re-parse AST; pass Prettier the AST Babel has already created. (This is blocked by an upgrade to Babel 7. I've left TODO comments with more information inside of `compile.js`.)
* #1313: Lazy-load Babel standalone from CDN as well to better support loading different versions of Babel.

DanBuild commented:
Deploy preview ready!

Built with commit 66828bfaca84c9df0fefced9b4570e56103bc407

https://deploy-preview-1297--babel.netlify.com

Daniel15 commented on js/repl/.gitignore:
```diff
@@ -0,0 +1,21 @@
+
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
```
Can we merge this into the root `.gitignore`?

Daniel15 commented on js/repl/package.json:
```diff
@@ -0,0 +1,30 @@
+  "name": "babel-repl",
+  "version": "0.1.0",
+  "private": true,
+  "homepage": "https://bvaughn.github.io/babel-repl",
```
Nit: Remove

Daniel15 commented on js/repl/public/index.html:
```diff
@@ -0,0 +1,52 @@
+<!doctype html>
```
I think that ideally this should use the Babel site styling rather than being a totally separate HTML page. We probably want the Babel top bar at least, for consistency.

bvaughn replied:
Agreed! This is just detritus from my demo site. I haven't yet looked into how to share HTML with the main babel site. 😄

Daniel15 replied:
@bvaughn  - Should be pretty straightforward. You just need to make a HTML file with "YAML front matter" so that Jekyll processes it. See how the old REPL does it: https://github.com/babel/babel.github.io/blob/master/repl.html

bvaughn replied:
Interesting. So the only tricky part is making it work nicely with dev workflow. (Adding front matter would break the `yarn dev` workflow for CRA.) Maybe this is a good enough reason to bail on CRA, although I like the idea of being able to test the REPL without runningt he whole (slow) Jekyll build.

Daniel15 replied:
In theory, the JS can be rebuilt entirely separately from the Jekyll build, if you can configure it to build it directly into the `_site` directory. That way, you'd only care about the Jekyll build if you touch the HTML. I wonder if Jekyll would get angry at us for touching the `_site` files directly though. Could try it and see 😛 

Yeah, Jekyll is slow... Maybe we should transition the site to something like Hugo, it's a lot faster.

Daniel15 replied:
In theory the build process for the JS could be whatever you wanted it to be, as long as the files end up in `_site`.

Daniel15 replied:
In theory the build process for the JS could be whatever you wanted it to be, as long as the files end up in `_site` somehow (either through Jekyll copying the files across, or through something else manually doing it)

bvaughn replied:
> In theory, the JS can be rebuilt entirely separately from the Jekyll build, if you can configure it to build it directly into the `_site` directory.

That's what the PR currently does (via `yarn build`) although nothing hooks into the `make build` process to run it.

> Yeah, Jekyll is slow...

I've been waiting for `npm start` to run for about 5 minutes and I'm stuck at "_Generating..._" 🤔

Daniel15 replied:
Wow, it shouldn't take that long... I think it normally takes ~5-10 seconds for me (which is slow, but bearable). I wonder if it's hung on something.

bvaughn replied:
Seems likely, although I `rm -rf ./_site && npm start` and it still seems to run to that point and then stick.

bvaughn replied:
Running `bundle exec jekyll serve` directly (with the `--verbose` flag) shows it getting hung after:
```
   GitHub Metadata: Initializing...
        Generating: Jekyll::GitHubMetadata::GHPMetadataGenerator finished in 0.001932 seconds.
        Generating: JekyllOptionalFrontMatter::Generator finished in 0.091768 seconds.
        Generating: JekyllReadmeIndex::Generator finished in 24.729063 seconds.
```
This is true on `master` as well. Hm. I wonder what broke. This worked for me at one point.

bvaughn replied:
Running `bundle exec jekyll serve` directly (with the `--verbose` flag) shows it getting hung after:
```
   GitHub Metadata: Initializing...
        Generating: Jekyll::GitHubMetadata::GHPMetadataGenerator finished in 0.001932 seconds.
        Generating: JekyllOptionalFrontMatter::Generator finished in 0.091768 seconds.
        Generating: JekyllReadmeIndex::Generator finished in 24.729063 seconds.
```
This is true on `master` as well. Hm. I wonder what broke. This worked for me at one point.

## Edit 1
A fresh clone also fails, but in a different way:

```sh
git clone git@github.com:babel/babel.github.io.git ./babel-new
cd babel-new
bundle install
npm i
npm start
```

Fails with:
```sh
> babel.github.io@1.0.0 start /Users/bvaughn/Documents/git/babel-new
> make serve

bundle check || bundle install; \
  bundle exec jekyll serve
The Gemfile's dependencies are satisfied
Configuration file: /Users/bvaughn/Documents/git/babel-new/_config.yml
Configuration file: /Users/bvaughn/Documents/git/babel-new/_config.yml
            Source: /Users/bvaughn/Documents/git/babel-new
       Destination: /Users/bvaughn/Documents/git/babel-new/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
  Liquid Exception: Could not locate the included file 'readmes/babel-core.md' in any of ["/Users/bvaughn/Documents/git/babel-new/_includes"]. Ensure it exists in one of those directories and, if it is a symlink, does not point outside your site source. in docs/usage/api.md
jekyll 3.4.3 | Error:  Could not locate the included file 'readmes/babel-core.md' in any of ["/Users/bvaughn/Documents/git/babel-new/_includes"]. Ensure it exists in one of those directories and, if it is a symlink, does not point outside your site source.
make: *** [serve] Error 1
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! babel.github.io@1.0.0 start: `make serve`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the babel.github.io@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bvaughn/.npm/_logs/2017-08-08T00_07_53_930Z-debug.log
```

## Edit 2
As far as the fresh checkout goes- looks like I needed to run `npm run build` before `npm start`. Didn't see that in the docs. Still having trouble with this checkout though. Hm.

## Edit 3
Checking out my branch in the new repo and `yarn build`ing this project before starting Jekyll up again caused it to hang also. I'm not familiar with Jekyll so I don't know how fragile it's supposed to be. Seems to be pretty easy to hang it though. Anyone with more Jekyll experience willing to lend a hand?

hzoo replied:
I'l just make a `make bootstrap` like babel's

Daniel15 commented on js/repl/public/manifest.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "React App",
+  "name": "Create React App Sample",
```
Not quite. 😛 

Do we need this manifest? What does it do?

bvaughn replied:
Whoops. Just create-react-app boilerplate. 😛

Daniel15 commented on js/repl/src/App.js:
```diff
@@ -0,0 +1,3 @@
+import Repl from './repl/Repl.js';
+
+export default Repl;
```
What's the purpose of just exporting `Repl`? Could we just use `repl/Repl.js` directly?

hzoo commented:
another feature we should port over later is handling console.log and friends

Daniel15 commented on js/repl/src/registerServiceWorker.js:
```diff
@@ -0,0 +1,108 @@
+// In production, we register a service worker to serve assets from local cache.
```
This seems like overkill given most of the assets come from a CDN. We can keep it if you feel it provides value though. What do you think?

existentialism replied:
It's from CRA, I'm fine with removing.

Daniel15 commented on js/repl/src/index.js:
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import './index.css';
+import App from './App';
```
Since this is just `repl/Repl.js`, just import that file directly.

Daniel15 commented on js/repl/src/repl/CodeMirrorPanel.js:
```diff
@@ -0,0 +1,63 @@
+    backgroundColor: '#FEE',
+    color: '#A00',
+    margin: 0,
+    padding: '0.25rem 0.5rem'
```
Could we just use a regular CSS file here to keep it consistent with the rest of the site? We could keep this CSS-in-JS thing if you like, but we should ensure that it extracts the CSS into `.css` files on deploy.

bvaughn replied:
I'm using Glamor everywhere within the REPL. Wouldn't it be more consistent to continue to use Glamor here?

Daniel15 replied:
Sure, we can use it if you like. I wonder if other people on the team have any thoughts about using Glamor / other CSS-in-JS technologies vs just using vanilla CSS.

existentialism replied:
I have no issues at all w/ using Glamor here (and the amount of styles seems easy enough to remove/replace later)

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+    super(props, context);
+
+    const persistedState = this._loadState();
+    console.log(
```
Remove?

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+      ...state,
+      ...this._compile(persistedState.code, state)
+    };
+    console.log(
```
Remove?

bvaughn commented:
"console.log and friends"? What do you mean? 😄

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+      lineWrap,
+      plugins,
+      presets
+    } = this.state;
```
Nit: This is useful for fields that are used multiple times, but I'm not sure how much value there is in using object spread for fields that are only used once. This adds 10 lines of code for little value - You could simply do `evaluate={this.state.evaluate}` in the JSX.

bvaughn replied:
This is just a habit, partly for consistency and partly for aesthetics, but I agree it's unnecessary. It's also the sort of thing I hate to have to think about while coding. 😄

bvaughn replied:
I think I've backed out the more gratuitous cases with 031fa4f

hzoo commented:
> Sorry was trying to have some fun, we have to capture the console and re-output it

https://github.com/babel/babel.github.io/blob/64c82a57d7e6552d8992ae54572ef82a7f97c426/scripts/repl.js#L593-L641

bvaughn commented:
It's not obvious to me at a glance why we do this?

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+  _presetsToArray(state: State = this.state): Array<string> {
+    const { presets } = state;
+
+    return Object.keys(presets)
```
If this is being used as a map, could we use a `Map` rather than an object? Given the latest React beta requires `Map` to be polyfilled, we can likely rely on it existing?

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+        const { code } = this.state;
+
+        this._checkForUnloadedPlugins();
+        this._updateCode(code);
```
Avoid the unnecessary object spread:

```
this._updateCode(this.state.code)
```

Daniel15 commented on js/repl/src/repl/Repl.js:
```diff
@@ -0,0 +1,327 @@
+  }, {});
+
+const styles = {
+  codeMirrorPanel: css({
```
What's the advantage of all this CSS-in-JS stuff?

Daniel15 commented on js/repl/src/repl/ReplOptions.js:
```diff
@@ -0,0 +1,196 @@
+      presetState,
+      toggleSetting
+    } = this.props;
+    console.log('<ReplOptions> lineWrap:', lineWrap, typeof lineWrap);
```
Remove?

Daniel15 commented on js/repl/src/repl/compile.js:
```diff
@@ -0,0 +1,57 @@
+    }).code;
+
+    if (config.prettify && window.prettier !== undefined) {
+      // TODO Don't re-parse; just pass Prettier the AST we already have.
```
👍

Daniel15 commented:
This is a super minor thing, but I think `js/repl/src/repl/` is a bit silly since `repl` is repeated. In keeping with the existing layout of the site files, I'd suggest using `css/repl/` for the CSS and `scripts/repl/` for the JavaScript (`scripts/repl.js` and `scripts/7.js` is the current REPL).

Let's make this accessible at `/replv2/` or `/repl_new/` or something while we iterate. 👍 

Fantastic work @bvaughn!

bvaughn commented:
Thanks @Daniel15!

Agreed that `repl/src/repl` is a bit silly. It didn't feel appropriate to put it under `js/` (since we may want to put other things there in the future?) I can certainly remove that nested `repl` folder though. There's no benefit in it.

> I'd suggest using `css/repl/` for the CSS and `scripts/repl/` for the JavaScript

That would require replacing CRA with a new build pipeline. I don't think I have the time/interest in doing that (at least not until...maybe...the upcoming weekend).

> Let's make this accessible at `/replv2/` or `/repl_new/` or something while we iterate. 👍

Running `yarn build` will sync the output to `_site/repl2` (based on @hzoo's comment on the issue). As the TODO section states though, this still needs to be wired up to `make build`.

hzoo commented:
Oh the console handling is just so you can just console output in the site without opening devtools

bvaughn commented:
Sorry for the clowny `console.log` statements I forgot to remove. Tried to tidy this up for PR this morning, last minute, between meetings and... 🤡  happens

existentialism commented on js/repl/src/PluginConfig.js:
```diff
@@ -0,0 +1,74 @@
+  {
+    label: 'es2015-loose',
+    package: 'babel-preset-es2015-loose',
+    version: '7'
```
Just for my own understanding, these version numbers here currently don't do anything since we pull it from `babel-standalone`, yea?

(Noticed this was 7 where the rest were 6)

bvaughn replied:
Ah. Yes. I wrote all of this intending to pull these from CDN too but then I realized standalone had them baked in. We could remove them.

existentialism replied:
Might still be useful re: babel-preset-env (at least until https://github.com/babel/babel-standalone/pull/90 is merged), we can mark the presets that `babel-standalone` includes.

bvaughn replied:
Yeah. That can be part of the TODO for whoever does that (maybe me, I dunno).

Daniel15 replied:
Even after https://github.com/babel/babel-standalone/pull/90 is merged, `babel-preset-env` will still be a separate package due to how large it is (as it needs all those compat lists to determine which plugins to execute). With all the other presets and plugins, they're small enough (compared to Babel itself) that there's not really an advantage of splitting them out into their own packages.

bvaughn replied:
Should be easy enough to accommodate this. We can add an `isPreloaded` attribute to `PluginConfig` and make the `version` attribute optional.

hzoo commented:
https://github.com/babel/babel.github.io/pull/1299 should make it easier! runs everything

bvaughn commented:
Cool. Thanks Henry.

As I mentioned yesterday, I could use a hand getting the Jekyll setup finished. I'll try to knock out the remaining "TODO" items (other than that) over the next couple of days.

hzoo commented:
Are you still having the same issue in https://github.com/babel/babel.github.io/pull/1297#discussion_r131804432? not sure what the issue is right now

bvaughn commented:
Yes. I have not looked into it further after mentioning in Slack last night that I'd prefer to spend my (limited) time finishing the non-Jekyll related TODO items.

Daniel15 commented:
No problem @bvaughn. No pressure at all. If you're having difficulties getting it working with Jekyll and don't want to spend more time on it, just let us know when you've completed all the work that you want to do and someone else can take over the work to get it working with Jekyll. 👍

Daniel15 commented:
For what it's worth, the Netlify build is simply doing `npm run build && jekyll build` on a fresh clone and that's working fine.

ritz078 commented:
@bvaughn Did to get prettier to work in safari ?

bvaughn commented:
@ritz078 No, Safari has a complaint with the Prettier bundle syntax. I didn't look into it.

bvaughn commented:
@Daniel15 Thanks! I've gotten side-tracked by ReactNative sync issues but I'll try to find time to wrap this up this weekend. (I likely won't spend much effort looking into Jekyll stuff though. I'll be happy to hand that off to someone with more experience.)

xtuc commented:
@ritz078 @bvaughn The issue with https://bundle.run/prettier@1.5.3 seems to come from Prettier. (issue in [bundle.run](https://github.com/Rich-Harris/packd/issues/25))

bvaughn commented:
Agreed! It looked like a Prettier+Safari issue which is why I didn't look into it. 😄

Daniel15 commented:
Given the fact that Prettier is optional + Safari has very low market share, maybe it's not worth looking into that at the moment 😛  Broken in Firefox is more of an issue though. We can leave Prettier integration until later. If it's just due to some missing transforms, it might be more reliable to manually bundle Prettier using Webpack as then we can tweak the Webpack settings to get it just right.

hzoo commented:
This is just because prettier targets Node 4, and a lot of tools are compiling down to that instead of node 0.10 so it uses ES6 syntax (unlike Babel where we compile to ES5).

ritz078 commented:
For a temporary fix I am using https://github.com/beautify-web/js-beautify instead of prettier. Though it also supports jsx, its not as good as prettier for JSX.

Daniel15 commented:
Shouldn't bundle.run compile it down to ES5 though? If not, what's the purpose of the service? It's the same with babel-standalone - Babel itself targets newer versions of Node, and babel-standalone runs additional Babel transforms so it can work in browsers too.

bvaughn commented:
Broken in Firefox _would_ be an issue- if it were broken in Firefox- but it works fine in Firefox for me. :smile: What makes you say that?

> it might be more reliable to manually bundle Prettier using Webpack as then we can tweak the Webpack settings to get it just right.

Prettier is huge. I think we should not do this.

hzoo commented:
Using prettier isn't a blocker at all, it's not a thing on the babel repl currently, and this is just an initial PR?

Daniel15 commented:
> Broken in Firefox would be an issue- if it were broken in Firefox- but it works fine in Firefox for me. 😄 What makes you say that?

Oh, I saw it in the issue that @xtuc linked to: https://github.com/Rich-Harris/packd/issues/25

> Prettier is huge. I think we should not do this.

How would it be different to loading it via bundle.run though? In the end, something somewhere is bundling it, and using a hosted service gives much less flexibility

Daniel15 commented:
In any case, let's shelve the Prettier support for now, and add it as a followup 😄

ritz078 commented:
@bvaughn I tried it in https://transform.now.sh/ and it was breaking in firefox (similar error to safari). We can load prettier on demand (in case other issues are fixed).

bvaughn commented:
I don't know what transform.now.sh is but Prettier works fine in the context of this REPL component for Firefox + Chrome + Android.

bvaughn commented:
> > Prettier is huge. I think we should not do this.
>
> How would it be different to loading it via bundle.run though? In the end, something somewhere is bundling it, and using a hosted service gives much less flexibility

Because I only load it _if_ someone clicks the "Prettier" option. (Same with Babili.) That's one of the nice things about the way the REPL is setup.

Daniel15 commented:
> Edit Unless you're proposing we still defer the loading of it, but lazy-load our own bundle- in which case, sure. 

Yeah, sorry, I should have been clearer that this is what I meant. Either make our own bundle, or push a `prettier-standalone` to npm that has a Webpack'd bundle and then load that via unpkg or CDNJS.

bvaughn commented:
Taking on ownership of publishing a build of Prettier doesn't sound that appealing to me 😄  I think in this case, since it's broken for Safari only- it's reasonable to leave it and let Prettier fix it but I'll defer to you.

hzoo commented:
Yeah don't worry about prettier, btw i'm ok with merging under a new url whenever so its easier to review/iterate over. Doesn't have to have all the functionality yet either

bvaughn commented:
Up to you, @hzoo. This isn't likely to conflict with anything, given that it's new code- but I'd also be fine if we want to merge and then iterate. I'll get some time (maybe tonight, but more likely over the weekend) to play with it a bit more.

bvaughn commented:
> For what it's worth, the Netlify build (which is what we use to deploy the production version of the site) is simply doing npm run build && jekyll build on a fresh clone and that's working fine.

FWIW

```sh
# Pulled master
# Rebased this branch
yarn install
yarn build
jekyll build
# Hangs where it hung before
```

Daniel15 commented:
I'm sorry you're hitting issues, @bvaughn :( Maybe try `bundle exec jekyll build` rather than just `jekyll build`. `bundle exec foo` is like the Ruby Bundler equivalent of doing `./node_modules/.bin/foo` - It runs the project's local version of Jekyll rather than whatever system-wide one you have installed.

I'm really confused as to why you're hitting issues... It's working fine for me, on both Windows and "Linux on Windows" 😕 
![](http://ss.dan.cx/2017/08/daniel%40Daniel-PC_mntcsrcbabel.github.io_10-20.44.27.png)

If you like, you could just focus on the React components themselves, and someone else (me or @hzoo or one of the other wonderful people in this thread) can do the Jekyll stuff. Alternatively, just create the directory structure that you *think* will work, rely on Netlify to build the site, and view the preview site to ensure the Jekyll build works 😛

hzoo commented:
Yeah relying on the netlify url seems easier than local actually lol - so fast? although I don't get https://deploy-preview-1297--babel.netlify.com/ because its redirecting on repl/ still?

Daniel15 commented:
The page isn't yet being built by Jekyll, so it won't be visible on the
preview site yet.

Sent from my phone.

On Aug 11, 2017 1:36 PM, "Henry Zhu" <notifications@github.com> wrote:

> Yeah relying on the netlify url seems easier than local actually lol - so
> fast? although I don't get https://deploy-preview-1297--babel.netlify.com/
> because its redirecting on repl/ still?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/babel/babel.github.io/pull/1297#issuecomment-321912836>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFnHUJIwsWilYbc6VwaqjD4JsQ8lDIJks5sXLtHgaJpZM4OwAuj>
> .
>

bvaughn commented:
I think this PR is ready for review 🎉

I've added a couple of follow-up items (explicitly listed in the PR summary) either because (a) I don't think we're quite ready to do them yet or (b) I don't think I'm the best person to do them.

I've updated the demo at [bvaughn.github.io/babel-repl](https://bvaughn.github.io/babel-repl/) to match the latest PR changes.

Daniel15 commented on js/repl/package.json:
```diff
@@ -0,0 +1,25 @@
+    "prettier": "^1.5.3",
+    "react": "^16.0.0-beta.3",
+    "react-dom": "^16.0.0-beta.3",
+    "react-scripts": "1.0.10"
```
Maybe merge this with the root `package.json` so we don't have to `yarn install` in two separate directories? I'm not sure what the benefit of splitting them is.

Daniel15 commented on js/repl/package.json:
```diff
@@ -0,0 +1,25 @@
+    "eject": "react-scripts eject",
+    "flow": "flow",
+    "postbuild": "cp -r ./build ../../_site/repl2",
+    "prettier": "prettier --single-quote --write 'src/**/*.{js,jsx}'",
```
Likewise this would be good for the entire site.

Daniel15 replied:
Likewise, this would be good to have for the entire site, not just for the REPL.

Daniel15 replied:
(I don't know why this comment appears twice)

Daniel15 commented on js/repl/public/index.html:
```diff
@@ -0,0 +1,55 @@
+      homescreen on Android. See https://developers.google.com/web/fundamentals/engage-and-retain/web-app-manifest/
+    -->
+    <link rel="manifest" href="%PUBLIC_URL%/manifest.json">
+    <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">
```
I guess we'll remove this `%PUBLIC_URL%` once the build of this page is moved into Jekyll.

bvaughn replied:
Yeah, this HTML will be replaced by one with Jekyll frontmatter by whoever does that integration.

Daniel15 commented on js/repl/src/CodeMirror.js:
```diff
@@ -0,0 +1,132 @@
+  keyMap: 'sublime',
+  matchBrackets: true,
+  mode: 'text/jsx',
+  lineNumbers: true,
```
Nit: Alphabetize? This list of options is *almost* alphabetical, except for `lineNumbers` being in the wrong spot. 😛

bvaughn replied:
Whoops.

Daniel15 commented on js/repl/src/Repl.js:
```diff
@@ -0,0 +1,413 @@
+
+    this.state = {
+      ...state,
+      ...this._compile(persistedState.code, state)
```
Just use object spread in the object above?
```
const state = {
  builtIns: persistedState.builtIns,
  // ......
  ...this._compile(.....)
};
```

bvaughn replied:
Trying to remember the reason this evolved the way it did. I'm sure it made sense while I was iterating. 🤔

bvaughn replied:
Ah, right. `_compile` requires params from `state`

Daniel15 replied:
Ohhh, I see now! Silly me.

bvaughn replied:
Not silly 😅 I forgot why I'd done it this way too. So the comment will help future-me.

Daniel15 commented on js/repl/src/PresetLoadingAnimation.js:
```diff
@@ -0,0 +1,56 @@
+});
+
+const styles = {
+  loadingAnimation: css({
```
Are all these styles extracted into real CSS files on build?

bvaughn replied:
They're inserted by JavaScript using [`insertRule`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/insertRule), why?

Daniel15 replied:
Ah, that's fine too. As long as it's not inline styles since they're not great for perf.

bvaughn replied:
[Glamor](https://github.com/threepointone/glamor) is really performant. I'm using it for the new React website too.

bvaughn replied:
Plus @threepointone works with us now and he's really helpful. 😁

Daniel15 replied:
The reason that I was asking was that some other solutions need you to extract the CSS into a regular CSS file for best performance. See this benchmark from Emotion, for example: https://github.com/tkh44/emotion/blob/master/docs/benchmarks.md. I wasn't sure if Glamor has the same perf characteristics.

>  I'm using it for the new React website too.
Is the new React website using server rendering with CSS-in-JS? That's an interesting approach, I haven't seen anyone do that before 😛  Or does the site require JS just to render static pages?

bvaughn replied:
The new React site is built with Gatsby, which does SSR, and uses the [Glamor plug-in](https://www.gatsbyjs.org/packages/gatsby-plugin-glamor/) which is optimized for the server-rendering.

Daniel15 commented on js/repl/src/ReplOptions.js:
```diff
@@ -0,0 +1,521 @@
+  presetState: PluginStateMap,
+  runtimePolyfillConfig: PluginConfig,
+  runtimePolyfillState: PluginState,
+  toggleEnvPresetSetting: ToggleEnvPresetSetting,
```
Maybe name all the callbacks with an "on" prefix? Up to you though.

bvaughn replied:
I don't feel strongly about this so 👍

Daniel15 commented on js/repl/src/Svg.js:
```diff
@@ -0,0 +1,26 @@
+
+type Props = {
+  className: string,
+  path: string
```
Maybe call this component `SvgPath` since it's specifically for rendering a path?

bvaughn replied:
That's pretty strongly implied, no? It's a property of an element named `Svg` 😁

Daniel15 replied:
Fair enough. I just thought it might get confusing.

Daniel15 commented on js/repl/src/Repl.js:
```diff
@@ -0,0 +1,413 @@
+        // Just evaluate the most recently compiled code.
+        try {
+          // eslint-disable-next-line
+          eval(this.state.compiled);
```
Shouldn't this only happen if evaluation is enabled? I saw a `config.evaluate` check for the other `eval` call.

bvaughn replied:
We'll only hit these lines of "evaluate" is checked. (It toggles `runtimePolyfillState.isEnabled`. The comments a few lines up mention this.)

Daniel15 replied:
Oops, thanks for the correction :)

Daniel15 commented on js/repl/src/index.css:
```diff
@@ -0,0 +1,24 @@
+  margin: 0;
+  padding: 0;
+  font-family: sans-serif;
+}
```
I guess this body styling can be removed once we merge the REPL into the site itself?

bvaughn replied:
Yes. Correct! The only bit I'd want to keep (probably) would be

```cs
box-sizing: content-box
```

bvaughn commented:
> This looks pretty solid to me. I left a few comments inline but they're all very minor. I couldn't comment on favicon.ico as it's a binary file, ut we can probably avoid the favicon.ico file and just use the root favicon.

Sure. This was just part of a standard create-react-app boilerplate. Once we use the Jekyll partial template thingy, the favicon would be the main Babel one anyway.

> With regards to integrating it into Jekyll, would you like someone to fork your fork and send you a pull request for it (into your repl-2 branch) to get it into this pull request itself, or would you prefer to merge this and iterate?

Either is fine! 😄

> One other question - Would it take much work to lazy load babel-standalone itself? In #1284 I changed Babel to be lazy loaded so that we could load a custom version (either from a CircleCI build, or a previous version from unpkg). I can work on integrating those changes into your version too, just wondering how difficult it'd be.

Not difficult. But probably okay to do as a follow-up item? (I don't mind doing it, just trying to prevent this PR from growing too much more.)

Daniel15 commented:
> But probably okay to do as a follow-up item?

Totally fine!

bvaughn commented:
>> But probably okay to do as a follow-up item?
>
> Totally fine!

We can turn the follow-up items (including this) into issues in [the "new repl" milestone](https://github.com/babel/website/milestone/1) @hzoo created.

bvaughn commented:
I've created tickets in the REPL milestone for all of the follow-up work.

bvaughn commented:
Thanks @existentialism!

bvaughn commented:
Thanks @xtuc!

I'm not sure if we should merge this yet or hold off for the Jekyll follow-up (#1309). Looks like CI is busted in master anyway so maybe it doesn't matter. I'll defer though since I'm still the noob here. 😄

Daniel15 commented:
The only issue with merging it now is that nobody is actually going to see it, since the HTML file is not yet accessible on the site 😛  Maybe that's okay for now though.

> Looks like CI is busted in master 

CI is only broken due to a lint error in one of the new blog posts. The actual website build is not broken.

bvaughn commented:
Okay! Here we go then! 🎉

Daniel15 commented:
Thanks @bvaughn! Since you were having issues with Jekyll, I'll see if I can get that part working. 👍

bvaughn commented:
<3 that's awesome, thanks!

On Aug 14, 2017 7:56 PM, "Daniel Lo Nigro" <notifications@github.com> wrote:

> Thanks @bvaughn <https://github.com/bvaughn>! Since you were having
> issues with Jekyll, I'll see if I can get that part working. 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/babel/website/pull/1297#issuecomment-322364384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABznbnLkE27yhHxZg0ABCqch-LACjK3ks5sYQjxgaJpZM4OwAuj>
> .
>

Daniel15 commented:
I worked out why you were having problems with Jekyll. The `js/repl/node_modules` directory is not in Jekyll's list of excluded directories, so it was trying to include all ~20,000 files in that directory as part of the website build 😛 

It took a while to track down, I only figured it out after watching exactly what the Ruby process was doing, and noticed that it was reading a large number of Markdown files from `node_modules`:
![](http://ss.dan.cx/2017/08/Process_Monitor_-_Sysinternals_www.sysinternals.c_14-21.24.29.png)

Anyways, I'll send a PR once I have something working.
